### ISSUE TITLE
Process completes but does not exit (leaving stranded workers)

### ISSUE THREAD

**Issue by smclinden** [log2timeline/plaso#399]
```
/usr/local/bin/log2timeline.py -p --vss-stores 1 srt.dump /media/usb1
```
```
plaso-1.3.1.post20151021

^CError in atexit._run_exitfuncs:
Traceback (most recent call last):
  File "/usr/lib/python2.7/atexit.py", line 24, in _run_exitfuncs
    func(*targs, **kargs)
  File "/usr/lib/python2.7/multiprocessing/util.py", line 325, in _exit_function
    p.join()
  File "/usr/lib/python2.7/multiprocessing/process.py", line 145, in join
    res = self._popen.wait(timeout)
  File "/usr/lib/python2.7/multiprocessing/forking.py", line 154, in wait
    return self.poll(0)
  File "/usr/lib/python2.7/multiprocessing/forking.py", line 135, in poll
    pid, sts = os.waitpid(self.pid, flag)
KeyboardInterrupt
Error in sys.exitfunc:
Traceback (most recent call last):
  File "/usr/lib/python2.7/atexit.py", line 24, in _run_exitfuncs
    func(*targs, **kargs)
  File "/usr/lib/python2.7/multiprocessing/util.py", line 325, in _exit_function
    p.join()
  File "/usr/lib/python2.7/multiprocessing/process.py", line 145, in join
    res = self._popen.wait(timeout)
  File "/usr/lib/python2.7/multiprocessing/forking.py", line 154, in wait
    return self.poll(0)
  File "/usr/lib/python2.7/multiprocessing/forking.py", line 135, in poll
    pid, sts = os.waitpid(self.pid, flag)
KeyboardInterrupt
Exception TypeError: "'NoneType' object is not callable" in <Finalize object, dead> ignored
Exception TypeError: "'NoneType' object is not callable" in <Finalize object, dead> ignored
Exception TypeError: "'NoneType' object is not callable" in <Finalize object, dead> ignored
Exception TypeError: "'NoneType' object is not callable" in <Finalize object, dead> ignored
Exception TypeError: "'NoneType' object is not callable" in <Finalize object, dead> ignored
Exception TypeError: "'NoneType' object is not callable" in <Finalize object, dead> ignored
Exception TypeError: "'NoneType' object is not callable" in <Finalize object, dead> ignored
Exception TypeError: "'NoneType' object is not callable" in <Finalize object, dead> ignored
Exception TypeError: "'NoneType' object is not callable" in <Finalize object, dead> ignored
Exception TypeError: "'NoneType' object is not callable" in <Finalize object, dead> ignored
Exception TypeError: "'NoneType' object is not callable" in <Finalize object, dead> ignored\


analyst   4035  3335  0 Oct29 pts/19   00:00:57 /usr/bin/python /usr/local/bin/log2timeline.py -p --vss-stores 1 srt.dump /media/usb1
analyst   4045  4035  0 Oct29 pts/19   00:01:02 /usr/bin/python /usr/local/bin/log2timeline.py -p --vss-stores 1 srt.dump /media/usb1
analyst   4048  4035  0 Oct29 pts/19   00:00:39 /usr/bin/python /usr/local/bin/log2timeline.py -p --vss-stores 1 srt.dump /media/usb1
analyst   4051  4035  0 Oct29 pts/19   00:00:58 /usr/bin/python /usr/local/bin/log2timeline.py -p --vss-stores 1 srt.dump /media/usb1
analyst   4054  4035  0 Oct29 pts/19   00:00:41 /usr/bin/python /usr/local/bin/log2timeline.py -p --vss-stores 1 srt.dump /media/usb1
analyst   4060  4035  0 Oct29 pts/19   00:00:44 /usr/bin/python /usr/local/bin/log2timeline.py -p --vss-stores 1 srt.dump /media/usb1
analyst   4063  4035  0 Oct29 pts/19   00:00:40 /usr/bin/python /usr/local/bin/log2timeline.py -p --vss-stores 1 srt.dump /media/usb1
analyst   4066  4035  0 Oct29 pts/19   00:00:37 /usr/bin/python /usr/local/bin/log2timeline.py -p --vss-stores 1 srt.dump /media/usb1
analyst   4068  4035  0 Oct29 pts/19   00:00:40 /usr/bin/python /usr/local/bin/log2timeline.py -p --vss-stores 1 srt.dump /media/usb1
analyst   4070  4035  0 Oct29 pts/19   00:11:09 /usr/bin/python /usr/local/bin/log2timeline.py -p --vss-stores 1 srt.dump /media/usb1
analyst   4073  4035  0 Oct29 pts/19   00:00:37 /usr/bin/python /usr/local/bin/log2timeline.py -p --vss-stores 1 srt.dump /media/usb1
analyst   4076  4035  0 Oct29 pts/19   00:00:37 /usr/bin/python /usr/local/bin/log2timeline.py -p --vss-stores 1 srt.dump /media/usb1
analyst   4083  4035  1 Oct29 pts/19   00:14:40 [log2timeline.py] <defunct>
```

**Comment by Onager** [comment]
Thanks for the report - could you try running with debugging turned on (-d) and direct the output to a file (--logfile=<some_file>)?  We need a little more information to investigate this.

**Comment by smclinden** [comment]
Running, as you suggested, right now.

**Comment by smclinden** [comment]
The debug log is 467M (uncompressed). How do you want me to handle this?

**Comment by Onager** [comment]
Probably only the last few lines are relevant - "cut -1000 | less", see if it has useful things and then make a Gist (https://gist.github.com/) of it.

Alternatively, it should compress pretty well, so you could upload it to some cloud service and send me the link.

**Comment by smclinden** [comment]
Can you send a personal email to mclinden@informed.net? I have a Dropbox link but I'd rather it be private.

**Comment by Onager** [comment]
I've taken a look at the log, it looks like process completion is being triggered, but at least one worker is still going. Not sure why this is happening at the moment.

**Comment by Onager** [comment]
At  21:14:00, worker_07 starts hashing a file. It looks to be a database file, and is probably large
At  21:14:06, processing is still incomplete, as extraction is still in process.
At  21:14:07, processing is marked as completed, and Plaso starts shutting down.
At  21:14:08, the storagewriter gets a QueueAbort, and exits.
At  21:15:08, the main process tries to terminate worker_07
At  21:21:53, worker_07 finishes hashing the file it started hashing at 21:14, then processes the file. When it's done, it stops monitoring the process queue.

So it looks to me like the processing status might need a tweak to account for workers that spend a long time hashing a file. 

@smclinden if this is causing you problems, you can likely work around it by turning off hashing with --hashers none

**Comment by smclinden** [comment]
Thanks! That is what I will do. I don't really need the hashes, anyway.

**Comment by Onager** [comment]
This is an interesting case - it looks like:
*   All pathspecs had been read from the pathspec queue, and worker_07 was looking at the last one
*   All other workers had completed
*   All events from other workers had been written by the writer
*   Worker_07 took a very long time to hash the file, and wasn't producing any events or errors during that time (it looks like the image was being processed over USB, so I/O bandwidth may have been limited)

So it looks like this bug will be triggered when the last (or one of the last) files to be processed is very large. Increasing some timeouts would likely fix this issue, but it's not particularly elegant or robust.

**Comment by smclinden** [comment]
I've been thinking about it. Very large files and volatile files (like databases) probably don't need to be hashed at indexing time, or there might be an option to exclude files exceeding a given size or with a given extension. This could speed up the processing, especially when dealing with something like Windows Small Business Server which contains databases, including mailboxes, Active Directory, etc.

**Comment by butljon** [comment]
I believe that I have something similar.  My Python interpreter:
jonathan@bearsden:~$ python
Python 2.7.9 (default, Mar  1 2015, 12:57:24) 
[GCC 4.9.2] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>>

When I run following lines (only example program - not what I am actually trying to do):

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
from multiprocessing import Process, Queue
import os

CONST=2810
THREADS=8
#CONST=2128
#THREADS=9


def f(i):
    myList = []
    for j in range (0, CONST):
        myList.append(j)
    q.put(myList)
    print 'process id:', os.getpid(), 'finished put'

q=Queue()
p=[]
for i in range (0,THREADS):
    p.append(Process(target=f, args=(i,)))

for i in range (0,THREADS):
    p[i].start()

for i in range(0,THREADS):
    p[i].join()
    
cntr=0
while q.qsize() != 0:
    q.get()
    cntr=cntr+1

print cntr,'queue elements!'
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

with
CONST=2810
THREADS=8
or
CONST=2128
THREADS=9

it works but with

CONST=2811
THREADS=8
or
CONST=2129
THREADS=9
it just hangs (and I am left with a load of "Sl" processes).

Does anyone have any ideas.  Python bug?  Workarounds?
Thanks in advance,

**Comment by joachimmetz** [comment]
@butljon how is this related to plaso?

**Comment by butljon** [comment]
In that Google noticed that both cases yield similar / the same error messages / traceback.  KeyboardInterrupt caused my program to dump:
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Traceback (most recent call last):
  File "/usr/lib/python2.7/multiprocessing/process.py", line 258, in _bootstrap
    self.run()
  File "/usr/lib/python2.7/multiprocessing/process.py", line 114, in run
    self._target(*self._args, **self._kwargs)
  File "User/programs/python/multi/multi_2.py", line 14, in f
Error in atexit._run_exitfuncs:
Traceback (most recent call last):
  File "/usr/lib/python2.7/atexit.py", line 24, in _run_exitfuncs
    func(*targs, **kargs)
  File "/usr/lib/python2.7/multiprocessing/util.py", line 325, in _exit_function
    p.join()
  File "/usr/lib/python2.7/multiprocessing/process.py", line 145, in join
    res = self._popen.wait(timeout)
  File "/usr/lib/python2.7/multiprocessing/forking.py", line 154, in wait
    return self.poll(0)
  File "/usr/lib/python2.7/multiprocessing/forking.py", line 135, in poll
    pid, sts = os.waitpid(self.pid, flag)
KeyboardInterrupt
Error in sys.exitfunc:
Traceback (most recent call last):
  File "/usr/lib/python2.7/atexit.py", line 24, in _run_exitfuncs
    func(*targs, **kargs)
  File "/usr/lib/python2.7/multiprocessing/util.py", line 325, in _exit_function
    p.join()
  File "/usr/lib/python2.7/multiprocessing/process.py", line 145, in join
    res = self._popen.wait(timeout)
  File "/usr/lib/python2.7/multiprocessing/forking.py", line 154, in wait
    return self.poll(0)
  File "/usr/lib/python2.7/multiprocessing/forking.py", line 135, in poll
    pid, sts = os.waitpid(self.pid, flag)

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

**Comment by Onager** [comment]
That's just what happens when you terminate a multiprocessing process using ctrl-c, it doesn't relate to Plaso.

**Comment by butljon** [comment]
Thought so, but OK - I'll look elsewhere.  Thanks again,

**Comment by joachimmetz** [comment]
@butljon have a look at http://stackoverflow.com/ might be a better place to ask your question.

**Comment by butljon** [comment]
thank you, Joachim

**Comment by butljon** [comment]
thanks again for the tip ... and there I found it!  wrong in my program was that one needs to get all the stuff off the queue before joining the threads (otherwise threads may not be able to terminate):  http://stackoverflow.com/questions/31170788/python-multiprocessing-threads-dont-close-when-using-queue
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
This is most likely because of this documented quirk of multiprocessing.Queue:

Bear in mind that a process that has put items in a queue will wait before terminating until all the
buffered items are fed by the “feeder” thread to the underlying pipe. (The child process can call the
cancel_join_thread() method of the queue to avoid this behaviour.)

This means that whenever you use a queue you need to make sure that all items which have been
put on the queue will eventually be removed before the process is joined. Otherwise you cannot be
sure that processes which have put items on the queue will terminate. Remember also that
non-daemonic processes will be joined automatically.

Basically, you need to make sure you get() all the items from a Queue to guarantee that all the
processes which put something into that Queue will be able to exit.
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

**Comment by Onager** [comment]
I have some code that should fix this now: https://codereview.appspot.com/273570043/

I'd prefer to avoid having an arbitrary "only hash files smaller than this" filter, both because of the arbitrariness of the number, and because de-duping by hash can be extremely useful (imagine a large DVD ISO, backup file etc.)

**Comment by Onager** [comment]
@smclinden Could you check that this fixes the problem you were having?

**Comment by smclinden** [comment]
It exited with different errors. I'll need to look into it.

**Comment by Onager** [comment]
I noticed a few other situations where there could be problems with zeromq in situations where all workers are busy but unproductive for long periods. If you get a chance, please retest with plaso from HEAD.

I'm going remove this as a release blocker for now.

**Comment by joachimmetz** [comment]
Closing this issue, behavior of multi processing has changed due to: https://github.com/log2timeline/plaso/issues/338
